src/main/kotlin/net/ccbluex/liquidbounce/integration/theme/component/types/minimap/MinimapTextureAtlasManager.kt

```diff
@@ -40,15 +42,18 @@ private const val ATLAS_SIZE: Int = 64
  */
 private const val FULL_UPLOAD_THRESHOLD: Int = 15
 
+private const val MAX_ATLAS_POSITIONS: Int = ATLAS_SIZE * ATLAS_SIZE - 1
+
 private val NOT_LOADED_ATLAS_POSITION = MinimapTextureAtlasManager.AtlasPosition(0, 0)
 
 class MinimapTextureAtlasManager {
     private val texture = NativeImageBackedTexture(ATLAS_SIZE * 16, ATLAS_SIZE * 16, false)
-    private val availableAtlasPositions = ArrayList<AtlasPosition>(ATLAS_SIZE * ATLAS_SIZE - 1)
+    private val availableAtlasPositions = ArrayList<AtlasPosition>(MAX_ATLAS_POSITIONS)
     private val dirtyAtlasPositions = hashSetOf<AtlasPosition>()
     private val chunkPosAtlasPosMap = hashMapOf<ChunkPos, AtlasPosition>()
 
     private val lock = ReentrantReadWriteLock()
+    private val semaphore = Semaphore(MAX_ATLAS_POSITIONS)
```

Comment:
Why is this needed when we have `availableAtlasPositions`

Reply:
The `removeLastOrNull` could be called when other threads(coroutines) are putting items back, so an exception will occur, we should wait until at least one item is am available

Reply:
But that would cause a thread to be stalled as soon as there is no more space in the Minimap texture atlas which is not a good method of handling failure :/

I will probably just add a proper fix haha

Reply:
> But that would cause a thread to be stalled as soon as there is no more space in the Minimap texture atlas which is not a good method of handling failure :/
> 
> I will probably just add a proper fix haha

I'll wait for it, I guess...

Reply:
I estimate that the scanning thread will be blocked for a few seconds at most, and the main thread will not be affected. Another choice is to change the functions (of Subscribers) to suspend functions...